src/internal/stream/merger.go

```diff
@@ -0,0 +1,79 @@
+package stream
+
+import (
+	"context"
+	"errors"
+	"io"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/stream/heap"
+)
+
+var _ Iterator[struct{}] = &Merger[struct{}]{}
+
+type mergeEntry[T any] struct {
+	it       Peekable[T]
+	priority int // lower is more important
+
+	peek T
+}
+
+type Merger[T any] struct {
+	its     []Peekable[T]
+	heap    heap.Heap[mergeEntry[T]]
+	isSetup bool
+}
+
+// NewMerger creates an iterator which merges the entries from its into a single iterator.
+// The entries will come out in ascending order.
+// The iterators slice can be thought of as layers, with higher layers masking the value of lower layers when the entries are equal.
+func NewMerger[T any](its []Peekable[T], lt func(a, b T) bool) *Merger[T] {
```

Comment:
This is a good interview question as well.

Reply:
Lol it took me more than an hour.  You can't copy, because the caller providing the destination means we can't do copy or assignment of values, the only way to write something is to call `Peek` or `Next`

Reply:
ahh, interesting.  i had to do this on the whiteboard at my google interview.  might have handwaved it as "oh it's just the merge from merge sort" tho.

Reply:
If we had gone with the nicer, but less performant `Next(context.Context) (T, error)` API, then I think it get's easier.  Maybe easy enough to squeeze into an interview.